### PR TITLE
new apply: on_destruct()

### PR DESCRIPTION
**Pull request by gesslar**
Notify object that it is scheduled for destruction

### CODE REVIEW

**Comment by thefallentree** [comment]
1.  the grammar autogen doesn’t seems to need to be changed: should revert that?
2. please add some test to the testsuite to make sure that cover error throwing case and non error throwing case
3.  In the current implementation it seems this will get called immediately when object is being destructed, and it’s going to be pretyy heavy : have you measured if this should be enabled per object instead? (Like set_heart_beat(1)” or something

**Comment by gesslar** [comment]
I can definitely do 1 and 2.

Adding this apply was really easy because it's just an apply. I don't think
it would be objectionable to have it enabled like set_heart_beat(), and I
can _try_ to do that, but I can't guarantee that I would be successful.

I just think this is really a missing feature because I have no idea when
things are being destructed outside of calling a remove() lfun, and things
do need to be cleaned up in those cases.

So, I will try.

Brian M. Workman
***@***.***

**Comment by gesslar** [comment]
@thefallentree I couldn't figure out a way to get rid of those two files. Maybe you can? But also, maybe this stuff should be in .gitignore 🤷‍♂️

But, tests and docs and stuff added, plus supporting efuns.

**Comment by gesslar** [comment]
Damnit, now my Local options are in e69a634b79ec17fc6a63fc2cee12890d6227fcac.

File: docs/apply/object/on_destruct.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+### SYNOPSIS
+
+    void on_destruct(void)
+
+### DESCRIPTION
+
+    The on_destruct() function is called by the driver when the object is
+    scheduled for destruction. This function is called before the object is
+    actually destructed, and is intended to allow the object to perform any
+    necessary cleanup before it is removed from the game.
```
Comment (michaelprograms) [inline docs/apply/object/on_destruct.md:20]: You should add:
```
### SEE ALSO

    set_notify_destruct(3), query_notify_destruct(3)
```

File: docs/efun/objects/query_notify_destruct.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    Objects do not receive this notification by default. To receive the call to
+    on_destruct(), the object must call set_notify_destruct(1) at some point
+    during its lifetime.
+
+    The query_notify_destruct() efun may only be called on any object.
+
+### SEE ALSO
+
+    set_notify_destruct, on_destruct, destruct
```
Comment (michaelprograms) [inline docs/efun/objects/query_notify_destruct.md:34]: Small nitpick: `set_notify_destruct(3), on_destruct(4), destruct(3)`

File: docs/efun/objects/set_notify_destruct.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    Objects do not receive this notification by default. To receive the call to
+    on_destruct(), the object must call set_notify_destruct(1) at some point
+    during its lifetime.
+
+    The set_notify_destruct() efun may only be called from within the object
+    that would like to receive the notification.
+
+### SEE ALSO
+
+    query_notify_destruct, on_destruct, destruct
```
Comment (michaelprograms) [inline docs/efun/objects/set_notify_destruct.md:31]: Same small nitpick as for query_notify_destruct.md

**Comment by gesslar** [comment]
Okay, I've reverted all of the bits that had been introduced accidentally. This is ready for a review.

**Comment by gesslar** [comment]
Okay! all checks passed!

**Comment by thefallentree** [comment]
please make the tests write a file or call a method , so that the tests actually fails if on_destruct was not called.

**Pull request merged**